pkg/pillar/cmd/agent.go

```diff
@@ -0,0 +1,64 @@
+package cmd
+
+import (
+	"flag"
+	"github.com/lf-edge/eve/pkg/pillar/agentlog"
+	"github.com/lf-edge/eve/pkg/pillar/pidfile"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// AgentBase - Interface for all agents to use the AgentBase.
+// Each agent is expected to have a Context that implements
+// this interface.
+type AgentBaseIntf interface {
+	AgentBaseContext() *AgentBaseContext
+	AddAgentSpecificCommandLineFlags()
+	ProcessAgentSpecificCLIFlags()
+}
+
+// AgentBaseCLIParams - stores all the common cli params
+type AgentBaseCLIParams struct {
+	DebugOverride bool
+}
+
+// AgentBaseContext - a struct that represents a general agent context
+type AgentBaseContext struct {
+	CLIParams   AgentBaseCLIParams
+	ErrorTime   time.Duration
+	WarningTime time.Duration
+	AgentName   string
+}
+
+// processCLIFlags - Add Flags common to all agents
+func processCLIFlags(agentBaseIntf AgentBaseIntf) {
+	debugPtr := flag.Bool("d", false, "Debug flag")
+	agentBaseIntf.AddAgentSpecificCommandLineFlags()
+	flag.Parse()
+	ctx := agentBaseIntf.AgentBaseContext()
+	ctx.CLIParams.DebugOverride = *debugPtr
+	agentBaseIntf.ProcessAgentSpecificCLIFlags()
+}
+
+// Run - a general run function that will handle all of the common code for agents
+func Run(agentSpecificContext AgentBaseIntf) {
+	ctx := agentSpecificContext.AgentBaseContext()
+	processCLIFlags(agentSpecificContext)
```

Comment:
I find this confusing. `AgentBase` interface  is passed to `Run()`, which makes sense. It then calls `AgentBase.AgentBaseContext()`, which is passed to `processCLIFlags()`, but the arg to `processCLIFlags()` is... `AgentBase`? Is that recursive? It is hard to understand.

Reply:
It does not pass AgentBase.AgentBaseContext() to processCLIFlags. That is just called so we have a ctx reference for later. I'll put this in the line right under to avoid confusion.

Reply:
I see how this might be confusing as the variable is called agentSpecificContext, but I called it that because the context of each agent will implement this interface. The changes to nodeagent show how that works.